src/Language/PureScript/Errors.hs

```diff
@@ -800,14 +801,40 @@ prettyPrintSingleError (PPEOptions codeColor full level showDocs relPath) e = fl
                       zipWith (Box.<>)
                       (Box.alignHoriz Box.left longestId <$> idBoxes)
                       tyBoxes
+              (xs, ys) = partition resultVariablesInConstraints idents
+              makeCodeBox = markCodeBox . indent . formatTS . unzip . take maxTSResults
             in [ line "You could substitute the hole with one of these values:"
-               , markCodeBox (indent (formatTS (unzip (take maxTSResults idents))))
+               , makeCodeBox xs
+               , line "You may also want these, but these may be less useful:"
+               , makeCodeBox ys
                ]
           _ -> []
       in
         paras $ [ line $ "Hole '" <> markCode name <> "' has the inferred type "
                 , markCodeBox (indent (typeAsBox ty))
                 ] ++ tsResult ++ renderContext ctx
+          where
+            resultVariablesInConstraints (_, t) =
+              not . S.null $ S.intersection constrained results
+              where
+                constrained = S.fromList (getConstrainedVariables t)
+                results = S.fromList (usedTypeVariables . last $ splitFunctionType t)
```

Comment:
Ideally, we should collect all three sets in one pass: constrained variables, variables appearing on the left, and variables appearing on the right.